cleanup opt-rocm references when using spack

srekolam opened the pull request:
when using spack and when building the hipblas and hipsparse ,it is found that  inside the CMakeCache.txt. there are some /opt/rocm references whenever there is rocm installation. The changes in the PR will address these issues.
for hipblas -it was showing up as below-
ROCM_DIR:PATH=/opt/rocm/share/rocm/cmake
HIP_run_hipcc:INTERNAL=/opt/rocm/hip/cmake/FindHIP/run_hipcc.cmake
HIP_run_make2cmake:INTERNAL=/opt/rocm/hip/cmake/FindHIP/run_make2cmake.cmake
and for hipsparse-
HIP_run_hipcc:INTERNAL=/opt/rocm/hip/cmake/FindHIP/run_hipcc.cmake
HIP_run_make2cmake:INTERNAL=/opt/rocm/hip/cmake/FindHIP/run_make2cmake.cmake

srekolam committed:
cleanup opt-rocm references when using spack

srekolam committed:
cleanup opt-rocm references when using spack

haampie commented on var/spack/repos/builtin/packages/hipblas/package.py:
```diff
@@ -31,6 +31,9 @@ class Hipblas(CMakePackage):
         depends_on('comgr@' + ver, type='build', when='@' + ver)
+        depends_on('rocm-cmake@' + ver, type='build', when='@' + ver)
+
+    patch('hipblas-remove-opt-rocm-reference.patch', when='@4.2.0:')
```
This patch shouldn't be necessary; CMAKE_MODULE_PATH is appended after it searches for rocm-cmake modules, and afterwards the only module being looked for is HIP if USE_CUDA is enabled (which we don't do).

Adding the dependency on rocm-cmake should be enough, spack adds it to CMAKE_PREFIX_PATH, which has higher precedence than the hard-coded PATHS, see https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure

haampie commented on var/spack/repos/builtin/packages/hipsparse/package.py:
```diff
@@ -39,6 +39,7 @@ class Hipsparse(CMakePackage):
 
     patch('e79985dccde22d826aceb3badfc643a3227979d2.patch', when='@3.5.0')
     patch('530047af4a0f437dafc02f76b3a17e3b1536c7ec.patch', when='@3.5.0')
+    patch('hipsparse-remove-opt-rocm-reference.patch', when='@4.2.0:')
```
Can you share what system version got picked up?

Also here, it should pick up spack's hip.

srekolam replied:
i have attached the CMakeCache.txt with out the patch, i had a rocm-4.3.0 installed in the /opt/rocm path and when i build the hipblas and hipsparse recipes using spack, the build succeeds but the cmakecache.txt seems to refer to few files from /opt/rocm  for HIP_run_hipcc etc.

srekolam commented:
[CMakeCache_hipsparse.txt](https://github.com/spack/spack/files/6965327/CMakeCache_hipsparse.txt)
[CMakeCache_hipblas.txt](https://github.com/spack/spack/files/6965328/CMakeCache_hipblas.txt)

haampie commented:
Hm... you're right. It's doing `find_package(HIP REQUIRED)`, from the cmake docs:

> If the MODULE option is not specified in the above signature, CMake first searches for the package using Module mode. Then, if the package is not found, it searches again using Config mode. A user may set the variable CMAKE_FIND_PACKAGE_PREFER_CONFIG to TRUE to direct CMake first search using Config mode before falling back to Module mode.

and rocm has both a FindHIP.cmake module and hip-config.cmake config :hankey:.

In my case it's finding the config file cause I don't have the module file, and still it builds... So, I'm thinking they just don't use it really. Let me check.

haampie commented:
I think we might want to set CMAKE_MODULE_PATH in spack and give it the same value as CMAKE_PREFIX_PATH, instead of patching the sources. Let me try that

haampie commented:
Ok, there is no agreed upon standard location for FindX.cmake files, and often they are used only internally by the config file of the same package.

Instead, let's just set `CMAKE_MODULE_PATH` to `spec['hip'].prefix.cmake` for these particular packages, that should solve the issue.

haampie committed:
Set CMAKE_MODULE_PATH to locate FindHIP.cmake

srekolam committed:
Merge branch 'rocm_rel_4.2_br10' of https://github.com/srekolam/spack into rocm_rel_4.2_br10

srekolam commented:
i added self.define('CMAKE_MODULE_PATH', self.spec['hip'].prefix.cmake) in both the hipblas and hipsparse and it is now showing as below in the CMakeCache.txt . i think that will fix the issue with out the patch . i updated the branch now


HIP_run_hipcc:INTERNAL=/spack/opt/spack/linux-centos8-skylake_avx512/gcc-8.4.1/hip-4.2.0-a2jhesmeredqfvbcg2setektl6spgxbq/cmake/FindHIP/run_hipcc.cmake
//Location of
HIP_run_make2cmake:INTERNAL=/spack/opt/spack/linux-centos8-skylake_avx512/gcc-8.4.1/hip-4.2.0-a2jhesmeredqfvbcg2setektl6spgxbq/cmake/FindHIP/run_make2cmake.cmake

srekolam committed:
Merge branch 'develop' into rocm_rel_4.2_br10

haampie commented on var/spack/repos/builtin/packages/hipblas/package.py:
```diff
@@ -53,3 +56,4 @@ def cmake_args(self):
 
     def setup_build_environment(self, env):
         env.set('CXX', self.spec['hip'].hipcc)
+        env.set('HIP_PATH', self.spec['hip'].prefix)
```
```suggestion
```

This is already covered for by hip/package.py

srekolam committed:
fix review comments
